#Random Forest

In [112]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, confusion_matrix

#Training Data

In [113]:
#load the dataset
data=pd.read_csv("/content/Training set.csv")
data.head()

Unnamed: 0  Unnamed: 0.1  Monthly_Inhand_Salary   Accounts of user  \
0        5279          5279             4377.072500                10   
1         473           473             2184.576667                 8   
2        1355          1355             1837.084167                 3   
3        1546          1546             6867.951667                 5   
4        1323          1323             1089.858750                 4   

   Credit cards user have  Interest Rate  Num_of_Loan  Delay_from_due_date  \
0                       6             27            6                   35   
1                       6             32            7                   51   
2                       7              3            2                   14   
3                       4              9            1                    6   
4                       7              4            2                   12   

   Num_of_Delayed_Payment  Changed_Credit_Limit  Outstanding_Debt  \
0                      18                 16.66           2950.80   
1                      19                 25.17           2989.13   
2                       8                  3.37           1336.73   
3                      12                 10.36            625.32   
4                       5                 11.90           1439.09   

   Total_EMI_per_month  Amount_invested_monthly  Monthly_Balance  \
0           154.172426               301.593795       271.941029   
1           132.481128                71.016720       294.959819   
2            26.746701                31.027274       375.934441   
3            66.457710               176.267359       704.070097   
4            14.350146                50.840361       303.795367   

   Utlization Ratio  Credit Inquiries  Cumulative Balance  Cash Flow  \
0                76                15                4009       4251   
1                75                19                2425       1073   
2                94                 0                1731       3322   
3                86                 8                1984       1614   
4                93                 0                2765       3986   

  Credit Score  
0         Best  
1         Best  
2         Good  
3     Standard  
4         Good

In [114]:
data.shape

(6094, 19)

In [115]:
X = data.drop('Credit Score', axis=1).values
y = data['Credit Score'].values

In [116]:
#split the data into training and testing data
X_train, X_test, y_train, y_test=train_test_split(X,y,test_size=0.2,random_state=45)

fitting the model

In [117]:
model = RandomForestClassifier(n_estimators=15,criterion='entropy')
model.fit(X_train, y_train)

RandomForestClassifier(criterion='entropy', n_estimators=15)

In [118]:
#scaling the testing data
y_pred = model.predict(X_test)

In [119]:
model.score(X_train, y_train)

0.9989743589743589

In [120]:
model.score(X_test, y_test)

0.9368334700574241

In [121]:
#confusion matrix
confusion_matrix(y_test, y_pred)

array([[759,   3,  28],
       [  5, 132,  12],
       [ 23,   6, 251]])

In [122]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

        Best       0.96      0.96      0.96       790
        Good       0.94      0.89      0.91       149
    Standard       0.86      0.90      0.88       280

    accuracy                           0.94      1219
   macro avg       0.92      0.91      0.92      1219
weighted avg       0.94      0.94      0.94      1219



#Testing Data

In [123]:
#load the dataset
data1=pd.read_csv("/content/Test SET.csv")
data1.head()

Unnamed: 0  Unnamed: 0.1  Monthly_Inhand_Salary   Accounts of user  \
0        7531          7531             3520.621667                 7   
1        1914          1914             2455.479851                 3   
2        1282          1282              592.179167                 6   
3        5517          5517             4976.090000                 9   
4        3610          3610             1720.165833                 7   

   Credit cards user have  Interest Rate  Num_of_Loan  Delay_from_due_date  \
0                       6             28            9                   32   
1                       5              5            3                    9   
2                       5             29            3                   41   
3                       7             30            9                   26   
4                       9             32            4                   47   

   Num_of_Delayed_Payment  Changed_Credit_Limit  Outstanding_Debt  \
0                      16                 27.12           4838.26   
1                      16                  9.48            595.94   
2                      22                 -0.34           2688.12   
3                      15                 25.00           3680.62   
4                      19                  7.41           1457.54   

   Total_EMI_per_month  Amount_invested_monthly  Monthly_Balance  \
0           220.401498               120.909497       260.751172   
1           183.764870                48.938091       380.803605   
2             9.575855                46.471450       283.170612   
3           432.581034               298.322639        36.705328   
4            62.671262               100.880899       268.464422   

   Utlization Ratio  Credit Inquiries  Cumulative Balance  Cash Flow  
0                85                19                3630       3229  
1                80                 0                4401       1920  
2                73                 8                1688       4100  
3                71                15                4304        402  
4                89                10                1528        645

In [124]:
data1.shape

(1524, 18)

In [125]:
data1

Unnamed: 0  Unnamed: 0.1  Monthly_Inhand_Salary   Accounts of user  \
0           7531          7531             3520.621667                 7   
1           1914          1914             2455.479851                 3   
2           1282          1282              592.179167                 6   
3           5517          5517             4976.090000                 9   
4           3610          3610             1720.165833                 7   
...          ...           ...                     ...               ...   
1519        7258          7258            11418.936670                 4   
1520         280           280             2582.813333                 7   
1521        3285          3285             1637.780833                 2   
1522        5567          5567             4622.948333                 5   
1523          51            51             1512.361667                 6   

      Credit cards user have  Interest Rate  Num_of_Loan  Delay_from_due_date  \
0                          6             28            9                   32   
1                          5              5            3                    9   
2                          5             29            3                   41   
3                          7             30            9                   26   
4                          9             32            4                   47   
...                      ...            ...          ...                  ...   
1519                       4             18            2                   30   
1520                       3              5            1                   16   
1521                       5              3            2                    2   
1522                       7             15            4                   26   
1523                       7             17            5                   51   

      Num_of_Delayed_Payment  Changed_Credit_Limit  Outstanding_Debt  \
0                         16                 27.12           4838.26   
1                         16                  9.48            595.94   
2                         22                 -0.34           2688.12   
3                         15                 25.00           3680.62   
4                         19                  7.41           1457.54   
...                      ...                   ...               ...   
1519                      18                 11.19            695.12   
1520                      10                  0.32           1237.46   
1521                       2                  3.58            288.32   
1522                       5                 11.18            112.36   
1523                      18                  9.95           2430.21   

      Total_EMI_per_month  Amount_invested_monthly  Monthly_Balance  \
0              220.401498               120.909497       260.751172   
1              183.764870                48.938091       380.803605   
2                9.575855                46.471450       283.170612   
3              432.581034               298.322639        36.705328   
4               62.671262               100.880899       268.464422   
...                   ...                      ...              ...   
1519           206.557194               550.114615       645.221858   
1520            22.688178               182.033671       323.559484   
1521            30.628576               100.871837       292.277671   
1522           127.790700               184.219006       410.285127   
1523            49.572062                55.298213       306.365891   

      Utlization Ratio  Credit Inquiries  Cumulative Balance  Cash Flow  
0                   85                19                3630       3229  
1                   80                 0                4401       1920  
2                   73                 8                1688       4100  
3                   71                15                4304        402  
4                   89          

In [131]:
from collections import Counter

In [137]:
l=[] 
l.append('Best') 
for i in range(1,1524):   
  k=data1.head(i)   
  output = model.predict(k)   
  d=Counter(list(output))
  Keymax = max(zip(d.values(), d.keys()))[1]
  l.append(Keymax)

/usr/local/lib/python3.8/dist-packages/sklearn/base.py:443: UserWarning: X has feature names, but RandomForestClassifier was fitted without feature names
  warnings.warn(
/usr/local/lib/python3.8/dist-packages/sklearn/base.py:443: UserWarning: X has feature names, but RandomForestClassifier was fitted without feature names
  warnings.warn(
/usr/local/lib/python3.8/dist-packages/sklearn/base.py:443: UserWarning: X has feature names, but RandomForestClassifier was fitted without feature names
  warnings.warn(
/usr/local/lib/python3.8/dist-packages/sklearn/base.py:443: UserWarning: X has feature names, but RandomForestClassifier was fitted without feature names
  warnings.warn(
/usr/local/lib/python3.8/dist-packages/sklearn/base.py:443: UserWarning: X has feature names, but RandomForestClassifier was fitted without feature names
  warnings.warn(
/usr/local/lib/python3.8/dist-packages/sklearn/base.py:443: UserWarning: X has feature names, but RandomForestClassifier was fitted without featu

In [138]:
data1['Credit Score']=l

In [139]:
data1

Unnamed: 0  Unnamed: 0.1  Monthly_Inhand_Salary   Accounts of user  \
0           7531          7531             3520.621667                 7   
1           1914          1914             2455.479851                 3   
2           1282          1282              592.179167                 6   
3           5517          5517             4976.090000                 9   
4           3610          3610             1720.165833                 7   
...          ...           ...                     ...               ...   
1519        7258          7258            11418.936670                 4   
1520         280           280             2582.813333                 7   
1521        3285          3285             1637.780833                 2   
1522        5567          5567             4622.948333                 5   
1523          51            51             1512.361667                 6   

      Credit cards user have  Interest Rate  Num_of_Loan  Delay_from_due_date  \
0                          6             28            9                   32   
1                          5              5            3                    9   
2                          5             29            3                   41   
3                          7             30            9                   26   
4                          9             32            4                   47   
...                      ...            ...          ...                  ...   
1519                       4             18            2                   30   
1520                       3              5            1                   16   
1521                       5              3            2                    2   
1522                       7             15            4                   26   
1523                       7             17            5                   51   

      Num_of_Delayed_Payment  Changed_Credit_Limit  Outstanding_Debt  \
0                         16                 27.12           4838.26   
1                         16                  9.48            595.94   
2                         22                 -0.34           2688.12   
3                         15                 25.00           3680.62   
4                         19                  7.41           1457.54   
...                      ...                   ...               ...   
1519                      18                 11.19            695.12   
1520                      10                  0.32           1237.46   
1521                       2                  3.58            288.32   
1522                       5                 11.18            112.36   
1523                      18                  9.95           2430.21   

      Total_EMI_per_month  Amount_invested_monthly  Monthly_Balance  \
0              220.401498               120.909497       260.751172   
1              183.764870                48.938091       380.803605   
2                9.575855                46.471450       283.170612   
3              432.581034               298.322639        36.705328   
4               62.671262               100.880899       268.464422   
...                   ...                      ...              ...   
1519           206.557194               550.114615       645.221858   
1520            22.688178               182.033671       323.559484   
1521            30.628576               100.871837       292.277671   
1522           127.790700               184.219006       410.285127   
1523            49.572062                55.298213       306.365891   

      Utlization Ratio  Credit Inquiries  Cumulative Balance  Cash Flow  \
0                   85                19                3630       3229   
1                   80                 0                4401       1920   
2                   73                 8                1688       4100   
3                   71                15                4304        402   
4                   89     

In [140]:
data1.shape

(1524, 19)

In [143]:
X1 = data1.drop('Credit Score', axis=1).values
y1 = data1['Credit Score'].values

In [144]:
X_train1, X_test1, y_train1, y_test1=train_test_split(X1,y1,test_size=0.2,random_state=45)

In [145]:
model.fit(X_train1, y_train1)
y_pred1 = model.predict(X_test1)

              precision    recall  f1-score   support

        Best       1.00      1.00      1.00       304
    Standard       0.00      0.00      0.00         1

    accuracy                           1.00       305
   macro avg       0.50      0.50      0.50       305
weighted avg       0.99      1.00      1.00       305



/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [146]:
model.score(X_train1, y_train1)

0.9991796554552912

In [147]:
model.score(X_test1, y_test1)

0.9967213114754099

In [148]:
confusion_matrix(y_test1, y_pred1)

array([[304,   0],
       [  1,   0]])

In [149]:
print(classification_report(y_test1, y_pred1))

              precision    recall  f1-score   support

        Best       1.00      1.00      1.00       304
    Standard       0.00      0.00      0.00         1

    accuracy                           1.00       305
   macro avg       0.50      0.50      0.50       305
weighted avg       0.99      1.00      1.00       305



/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


#Conclusion

 TRAINING SET : The accuracy of training data is 0.998 and for the testing data is 0.936



 
TESTING SET   : The accuracy of training data is 0.999 and for the testing data is 0.996
